llvm/lib/Target/AMDGPU/AMDGPU.td

```diff
@@ -1426,7 +1425,6 @@ def FeatureISAVersion9_4_Common : FeatureSet<
   [FeatureGFX9,
    FeatureGFX90AInsts,
    FeatureGFX940Insts,
-   FeatureFmaMixInsts,
```

Comment:
...How did this happen? gfx906 replaced mad_mix with fma_mix, and then all the MI went back to mad_mix? 

Looking at the features lists, I'm not sure mad_mix is properly added. Can you confirm that we have test coverage for the mad_mix cases for 908/90a/940?

Reply:
I think we shall not change it. This is quite unfortunate mess in the asm, but that is what really happened in Vega 12 and then carried forward to later gfx9 based products (at least according to the documentation):

MAD_MIXLO_F16/MAD_MIXHI_F16/MAD_MIX_F32 used to be Mad f32 based before, and FMA based on all later chips.

The most important statement there: "Name not changed".

So it is the same encoding, but different instructions before and after gfx906. Unfortunately, SP3 decided to preserve the name. I do not think we shall follow.

That said, run lines for gfx908 can be added to mad-mix*.ll tests and I believe shall use fma_mix.

Reply:
That's what I thought, but looking at the manual there is no mention of FMA in any of the MI manuals. gfx906 did the ugly mad->fma encoding switch, but all of the 908/90a/940 manuals only talk about mad

Reply:
What I was quoting was from SPG, not ISA manuals. Then note that implementation is always done according to the SPG and ISA manuals come later.

Reply:
But the rename happened in 908, and then the rename got lost after that? I'm confused

Reply:
It is not lost, we are using fma_mix on all later gfx9. It would be lost after this change. But it is we have renamed it, not SP3.

Reply:
The problem was reported for GFX90a -- they noticed presence of v_fma_mix, which is not supposed to occur based on the ISA doc.

As far as GFX90a is concerned, I noticed that v_fma_mix (1) is not in the ISA doc (2) is not supported by sp3. On the other hand, v_mad_mix is the opposite. So it should be fixed, as this PR is trying to do?

Reply:
I don't think it shall be fixed. It is really fma, not mad. That is a bad naming choice in the ISA doc. In particular if fixed like that you will get wrong selection (check `SITargetLowering::isFPExtFoldable`).

Reply:
So you are saying that both the MI200 and MI300 ISA docs are wrong, and v_fma_mix should be a valid instruction for gfx908/90a/940?

Reply:
I am not saying they are wrong. I am saying it was wrong decision to change instructions and keep the name.

Reply:
> I am not saying they are wrong. I am saying it was wrong decision to change instructions and keep the name.

Could you please elaborate on this? (1) Is the "wrong decision" in the code or in the ISA doc? (2) By "change instructions and keep the name", did you mean the mad_mix  name is kept in the ISA doc but is actually used for a different instruction than the mad_mix instruction in previous versions, say gfx900?

Reply:
> > I am not saying they are wrong. I am saying it was wrong decision to change instructions and keep the name.
> 
> Could you please elaborate on this? (1) Is the "wrong decision" in the code or in the ISA doc?

In the ISA doc.

> (2) By "change instructions and keep the name", did you mean the mad_mix name is kept in the ISA doc but is actually used for a different instruction than the mad_mix instruction in previous versions, say gfx900?

Yes. And the selection code uses ISD::FMAD to match mad_mix while ISD::FMA for fma_mix.

Reply:
Ok, I guess the decision is not to fix.

Reply:
The actual codegen treatment appears to be correct. For SP3 compatibility we could accept the mad names on the targets where it's really an FMA, I suppose. Is there a reason to do it? 

The ISA made a grave error, and SP3 made it worse